Queue override via endpoint instance property

SzymonPobiega opened the pull request:
Connects to Particular/V6Launch#68

Refactors the override local address code so that now it uses a `queue` property on the endpoint instance class. 
Removes the `OverrideAddressTranslation` API from MSMQ
Adds a `AddEndpointInstances` API to MSMQ which can be used in conjunction with `WithQueue` instance extension method to define routing to instances that have overridden queue name.

TODO:
 * [ ] Remove the translation rule hack from routing samples

SzymonPobiega commented:
@bording @DavidBoike @timbussmann please take a look (context in the routing slack channel)

timbussmann commented on src/NServiceBus.AcceptanceTests/Routing/When_overriding_local_address.cs:
```diff
@@ -0,0 +1,76 @@
+    using NUnit.Framework;
+    using Conventions = NServiceBus.AcceptanceTesting.Customization.Conventions;
+
+    public class When_overriding_local_address : NServiceBusAcceptanceTest
```
how does that test differ from `When_overriding_input_queue_name`?

SzymonPobiega replied:
It uses send options to send instead of routing so can be executed for all transports. `When_overriding_input_queue_name` is MSMQ-specific because the code-first instance API are MSMQ-specific.

timbussmann commented on src/NServiceBus.Core/Routing/EndpointInstance.cs:
```diff
@@ -56,6 +56,15 @@ public EndpointInstance SetProperty(string key, string value)
+        /// Creates a new endpoint instance with specified discriminator.
+        /// </summary>
+        /// <param name="discriminator">New discriminator value.</param>
+        public EndpointInstance SetDisctiminator(string discriminator)
```
spelling error

timbussmann commented on src/NServiceBus.Core/Transports/Msmq/ConfiguredEndpointInstances.cs:
```diff
@@ -0,0 +1,23 @@
+    using Features;
+    using Routing;
+
+    class ConfiguredEndpointInstancesFeature : Feature
```
I'm not really understanding this feature. Why would you do that?

timbussmann commented on src/NServiceBus.Core/Transports/Msmq/EndpointInstanceExtensions.cs:
```diff
@@ -18,5 +18,17 @@ public static EndpointInstance AtMachine(this EndpointInstance instance, string
+        /// </summary>
+        /// <param name="instance">A plain instance.</param>
+        /// <param name="queueName">Queue name.</param>
+        public static EndpointInstance WithQueue(this EndpointInstance instance, string queueName)
```
I'm good with that, but this needs (acceptance) tests

SzymonPobiega replied:
`NServiceBus.Msmq.AcceptanceTests/When_overriding_address_translation` uses it.

timbussmann commented on src/NServiceBus.Core/Transports/Receiving.cs:
```diff
@@ -15,17 +15,24 @@ internal Receiving()
                 {
-                    var p = s.Get<TransportInfrastructure>().BindToLocalEndpoint(new EndpointInstance(baseQueueName, userDiscriminator));
-                    s.SetDefault("NServiceBus.EndpointSpecificQueue", transportInfrastructure.ToTransportAddress(new LogicalAddress(p)));
+                    mainInstance = mainInstance.SetProperty("queue", baseQueueName);
```
:-1: to leak msmq specific settings into the `Receiving` feature.

SzymonPobiega replied:
That's not MSMQ concept. I've tried limiting the implementation of `queue` to MSMQ only but it results in inconsistency where the current endpoint transport address is generated in a different way than other endpoints. Decided to move this here and just not expose it in different transports for the outgoing messages, only for the local endpoint.

This way we can keep the logical stuff well separated from the physical stuff.

timbussmann replied:
I don't think the Properties are a good way to expose core concepts on `EndpointInstance`. In that case it should be a property on EndpointInstance directly as you have to ensure that all transports know about this property when implementing `ToTransportAddress`.

Can't we keep routing and input queue setup seperated (use "queue" property for transport specific routing, use a core only concept to implement `OverrideLocalAddress`).

SzymonPobiega replied:
Fair enough. That's a bad stringly-typed API. 

We can limit that to MSMQ but that won't give us the benefits we wanted with regards to samples, at least not all of them.  Let me think about an alternative.

SzymonPobiega replied:
@timbussmann how about setting the `queue` property **and** passing the overridden queue name as the endpoint name to the transport? Transports that are not aware of `queue` will work just fine without any modifications.

SzymonPobiega commented:
@timbussmann I switched back to overriding the endpoint name in the `LocalAddress` when passing to transport. This way we don't need to touch the transport. I left the "queue" property as without it the `EndpointInstance` stored in settings would have incomplete information compared to `EndpointInstances` read e.g. from the instance mapping file.

bording commented on src/NServiceBus.Core/Routing/EndpointInstance.cs:
```diff
@@ -56,6 +56,24 @@ public EndpointInstance SetProperty(string key, string value)
+        /// Creates a new endpoint instance with specified discriminator.
+        /// </summary>
+        /// <param name="discriminator">New discriminator value.</param>
+        public EndpointInstance SetDiscriminator(string discriminator)
```
Both this and `SetEndpoint` really should be static factory methods since they return a new instance, and don't modify the existing one.

Looking at the rest of this class, `SetProperty` should be a factory method as well.

bording replied:
Alternatively, is there a reason these need to return new instances vs. updating the existing instance with the new values?

bording commented on src/NServiceBus.Msmq.AcceptanceTests/When_overriding_address_translation.cs:
```diff
@@ -3,14 +3,16 @@
     using NUnit.Framework;
 
-    public class When_overriding_address_translation : NServiceBusAcceptanceTest
+    public class When_overriding_local_address : NServiceBusAcceptanceTest
```
This class name has changed, but the file name hasn't been updated.

bording commented:
While I'm sure the current changes to the Receiving feature are functional, they feel very hacky at this point, and it's hard to follow what even is going on in there.

timbussmann commented on src/NServiceBus.Core/Transports/Msmq/MsmqConfigurationExtensions.cs:
```diff
@@ -62,14 +63,14 @@ public static InstanceMappingFileSettings InstanceMappingFile(this RoutingSettin
-        /// <param name="translationRule">New translation rule.</param>
-        public static TransportExtensions<MsmqTransport> OverrideAddressTranslation(this TransportExtensions<MsmqTransport> config, Func<LogicalAddress, string> translationRule)
+        /// <param name="instances">A list of endpoint instances.</param>
+        public static void AddEndpointInstances(this RoutingSettings<MsmqTransport> config, params EndpointInstance[] instances)
```
we essentially wanted to avoid hardcoding instances on the basic API and now we introduce this again?

timbussmann commented:
I'm probably missing something important here, but would the simplest approach be to support the `queue` property in msmq and then define the custom queue names in the instance mapping file? Yes this will require users to use the instance mapping file if they overwrite the local input queue but want to keep the routing on a logical level (otherwise they can just define the new queue name as the logical endpoint?)

SzymonPobiega commented:
@timbussmann the only problem with that is that when you send to an endpoint with overridden queue name, you would pass something like this to the transport:

```
new EndpointInstance("Destination").SetProperty("queue", "DestQueue")
```

while from the perspective of the "DestinationEndpoint" when it creates the address for its own receiver it would be

```
new EndpointInstance("DestQueue")
```

that instance would get added to the settings dictionary, potentially creating some misunderstanding since it lies about the name of the endpoint. Even if it does not break any existing code, it might break something in future when somebody expects this to be the logical name of the endpoint.

I am tying to figure out if there is even a way to fix this without affecting the transport seam which I would like to avoid if possible.

SzymonPobiega commented:
Continuing my train of thought, we can do following in the core:
 * In the settings, store the `EndpointInstance` value that has correct endpoint "Destination"
 * When passing to the `ToTransportAddress`, on the fly create a new `EndpointInstance` with overridden endpoint name "DestQueue" but not store it anywhere. This ensures no changes to other transports are required
 * In the `Bind` method MSMQ would add the `queue` prop to the endpoint instance to ensure that the instance object looks the same, no matter if you look at the sender or at the receiver side.

SzymonPobiega commented:
@timbussmann @bording I pushed a new version. I think the `Receiving` feature is not a bit more clean. We no longer have the offending `EndpointInstance` in the settings. It is replaced by a more appropriate `LogicalAddress`.

The compromise here is that the `LogicalAddress` wraps an endpoint instance instead of having a native `QueueName` property.

bording commented:
What scenario is this trying to solve? For the main acceptance test, it's not clear to me why the SendOptions are needed when you could instead add `routing.RouteToEndpoint(typeof(Message), ReceiverQueueName)` and the test passes fine, even on MSMQ.

Why do we have an MSMQ-specific version of this same test that uses an additional API, one that reintroduces a way to hard-code instances into the non-advanced API, which we intentionally removed?

Are we trying to let multiple instances of a single endpoint have unique queue names for each instance? If so, why would we want to support that?

bording commented on src/NServiceBus.Core/LogicalAddress.cs:
```diff
@@ -10,28 +10,66 @@
+        /// Creates an address scoped inside this one.
+        /// </summary>
+        /// <param name="qualifier">The qualifier for the new address.</param>
+        public LogicalAddress CreateSubscope(string qualifier)
```
Like the Set methods before, I think that this method and `CreateIndividualized` would be better if they were static methods where you passed in an existing instance. That makes it more obvious that you are creating a new instance that you need to store.

SzymonPobiega replied:
I think the `Create` prefix is descriptive enough. It is quite obvious here that this class is immutable. It does not differ from the `string` API in the way it uses instance factory methods.

SzymonPobiega commented:
Although `routing.RouteToEndpoint(typeof(Message), ReceiverQueueName)` would technically work in this case, it would not work in a real-life scenario when we need `@machine`. I wanted the test to be as close to real-life usage as possible.

That MSMQ-specific API has been added in place of the `SetAddressTranslationRule` method but I think you are right, we can live without it because we have filed-based instance mapping. I'll remove it.

The problem we're trying to solve here is routing to instances which have overridden queue names, even if they are part of a scaled-out endpoint.

bording commented:
> That MSMQ-specific API has been added in place of the SetAddressTranslationRule method but I think you are right, we can live without it because we have filed-based instance mapping. I'll remove it.

:+1: to this idea. Let's take advantage of the file to set the queue property.

> The problem we're trying to solve here is routing to instances which have overridden queue names, even if they are part of a scaled-out endpoint.

For the most part, shouldn't this be MSMQ-specific, then? Other transports would only have the logical endpoint, and if the queue has been changed, the senders should be able to just use that name as the endpoint, letting a regular `routing.RouteToEndpoint()` call be enough to handle this.

Or is there something I'm missing here?

Let's go ahead and remove the current MSMQ API and replace it with using the instance mapping file, and then we can take a look how this PR is shaping up.

SzymonPobiega commented:
@bording I removed these APIs. Please take a look again.

bording commented:
Made a few changes to clean up a few things.

Overall this is good, though I think `LogicalAddress` still needs some work. While I think I'd still prefer static methods, I'm OK with the current Create methods. I'm thinking `CreateQualifiedAddress` and CreateCreateIndividualizedAddress` would be better names for the new methods. The xml comments for that class probably need a revision pass also. It keeps mentioning "instance name" when its actually talking about an endpoint instance. The new constructor that has several parameters also mentions being provided an endpoint instance.

I'm also thinking it might make sense to make some changes to the instance mapping schema file, to add machine and queue properties in there directly, and get rid of the "anyAttribute". That would let us validate against those specific properties and catch typos.

SzymonPobiega commented:
@bording @andreasohlund I renamed the factory methods on `LogicalAddress` class and fixed the XML comments. Rebased and squashed. I think that the possible fixes in the routing file schema should go with a separate pull.

andreasohlund commented:
needs one more rebase :)

andreasohlund commented:
Would this effect any of the transports? (or other downstreams)

SzymonPobiega commented:
This one not because the `LogicalAddress` is the same. But I think about a follow up one which would clean up the `LogicalAddress` class and that one would need the downstream PRs (but very very simple ones).

andreasohlund commented:
LGTM, please merge if you're happy with this @bording
